### ISSUE TITLE
Explicit CoreML Conversion Pipeline

### ISSUE THREAD

**Issue by harper-lambda** [roboflow/rf-detr#318]
### Search before asking

- [x] I have searched the RF-DETR [issues](https://github.com/roboflow/rf-detr/issues) and found no similar feature requests.

### Description

I am interested in converting the PyTorch models into CoreML. I know there are several titled issues related to this matter, but those have all either led to no resolution or link to now unavailable PRs. I have tried a number of approaches, converting directly from PyTorch and using ONNX as an intermediary but to no avail.

It would seem that the issue lies in the bicubic upsampling function not having a direct conversion between PyTorch and CoreML amongst a number of other incompatibilities.

I saw in the docs that there is a pipeline running on the backend of Roboflow Universe, that converts the models to CoreML for iOS inference, so I assume it is indeed possible.

Is this a planned feature on the roadmap for this repo? If not, is there something that must be modified with the model layers or inference code to make it compatible?

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by landchenxuan** [comment]
Hi all, I built an open-source tool  (with the help of Claude Code) that converts RF-DETR v1.5.1 (both detection and segmentation) to CoreML: [rf-detr-to-coreml](https://github.com/landchenxuan/rf-detr-to-coreml)

Huge thanks to @timnielen — studying [your fork](https://github.com/timnielen/rf-detr) is what made me realize this was possible. The core insight that deformable attention needs to be reshaped to stay within CoreML's rank-5 limit came directly from your work.

My implementation takes a different approach — runtime monkey-patches instead of forking — so it works with unmodified rfdetr installed via pip. It also adds:

  - Segmentation support (all 6 seg variants)
  - FP32-only pipeline with detailed analysis of why FP16 fails catastrophically (grid_sample precision)
  - Sub-pixel accuracy: < 0.01 px box diff on confident detections, verified across 17 real images
  - 1.7-1.9x faster than PyTorch MPS on Apple Silicon (GPU acceleration)
  - Comparison showing [why direct CoreML beats ONNX→CoreML](https://github.com/landchenxuan/rf-detr-to-coreml#why-not-onnx--coreml) (single graph vs 13-97 partitions)

```
  git clone https://github.com/landchenxuan/rf-detr-to-coreml.git
  cd rf-dert-to-coreml
  pip install -e .
  rfdetr-coreml --model nano
```
Hope this helps anyone trying to run RF-DETR on Apple devices!

**Comment by DiegoWu** [comment]
> Hi all, I built an open-source tool (with the help of Claude Code) that converts RF-DETR v1.5.1 (both detection and segmentation) to CoreML: [rf-detr-to-coreml](https://github.com/landchenxuan/rf-detr-to-coreml)
> 
> Huge thanks to [@timnielen](https://github.com/timnielen) — studying [your fork](https://github.com/timnielen/rf-detr) is what made me realize this was possible. The core insight that deformable attention needs to be reshaped to stay within CoreML's rank-5 limit came directly from your work.
> 
> My implementation takes a different approach — runtime monkey-patches instead of forking — so it works with unmodified rfdetr installed via pip. It also adds:
> 
> * Segmentation support (all 6 seg variants)
> * FP32-only pipeline with detailed analysis of why FP16 fails catastrophically (grid_sample precision)
> * Sub-pixel accuracy: < 0.01 px box diff on confident detections, verified across 17 real images
> * 1.7-1.9x faster than PyTorch MPS on Apple Silicon (GPU acceleration)
> * Comparison showing [why direct CoreML beats ONNX→CoreML](https://github.com/landchenxuan/rf-detr-to-coreml#why-not-onnx--coreml) (single graph vs 13-97 partitions)
> 
> ```
>   git clone https://github.com/landchenxuan/rf-detr-to-coreml.git
>   cd rf-dert-to-coreml
>   pip install -e .
>   rfdetr-coreml --model nano
> ```
> 
> Hope this helps anyone trying to run RF-DETR on Apple devices!

Hey have you tried benchmarking this on IOS devices? I used your converter for a fine-tuned RF-DETR nano model and got around 61 ms latency on iPhone 14 Pro, which is quite unacceptable for my use case (live video so 30+ fps😢
I profiled the model and like you described, I am seeing only GPU is used and no ANE support. Do you have updates on making the FP 16 export works with ANE?

**Comment by landchenxuan** [comment]
Hi @harper-lambda, FP16 export is now supported; please validate it with your own model, data, thresholds, and inference pipeline before using it in production.
